File: DSA-lab/modules/01-komplexitaetsanalyse-big-o/projects/01-basic/runtime_lab.py
"""Benchmark five functions with growth from constant to quadratic.

The curve labels are intentionally neutral. Learners should derive the
complexity classes from the implementations before inspecting the measurements.
"""

from __future__ import annotations

import argparse
import csv
import math
from dataclasses import dataclass
from pathlib import Path
from statistics import median
from time import perf_counter
from typing import Callable, Mapping, Sequence

WorkFunction = Callable[[int], int]


def curve_a(size: int) -> int:
    """Perform a fixed amount of work independent of size."""
    operations = 0
    for _ in range(32):
        operations += 1
    return operations


def curve_b(size: int) -> int:
    """Repeatedly halve the remaining problem size."""
    operations = 0
    remaining = size
    while remaining > 1:
        for _ in range(32):
            operations += 1
        remaining //= 2
    return operations


def curve_c(size: int) -> int:
    """Visit one item for every unit of input."""
    operations = 0
    for _ in range(size):
        for _ in range(8):
            operations += 1
    return operations


def curve_d(size: int) -> int:
    """Perform a halving process for every input item."""
    operations = 0
    for _ in range(size):
        remaining = size
        while remaining > 1:
            operations += 1
            remaining //= 2
    return operations


def curve_e(size: int) -> int:
    """Visit every unordered pair of distinct positions."""
    operations = 0
    for left in range(size):
        for _ in range(left):
            operations += 1
    return operations


FUNCTIONS: Mapping[str, WorkFunction] = {
    "A": curve_a,
    "B": curve_b,
    "C": curve_c,
    "D": curve_d,
    "E": curve_e,
}

DEFAULT_SIZES = (128, 256, 512, 1_024, 2_048, 4_096)
DEFAULT_OUTPUT_DIR = Path(__file__).resolve().parent / "results"


@dataclass(frozen=True)
class BenchmarkConfig:
    """Configuration for a complete benchmark series."""

    sizes: tuple[int, ...] = DEFAULT_SIZES
    repeats: int = 5
    min_sample_seconds: float = 0.01
    max_iterations: int = 1_048_576

    def validate(self) -> None:
        """Raise ValueError if the configuration cannot produce valid data."""
        if not self.sizes:
            raise ValueError("at least one input size is required")
        if any(size < 2 for size in self.sizes):
            raise ValueError("all input sizes must be at least 2")
        if any(left >= right for left, right in zip(self.sizes, self.sizes[1:])):
            raise ValueError("input sizes must be strictly increasing")
        if self.repeats < 1:
            raise ValueError("repeats must be positive")
        if self.min_sample_seconds <= 0:
            raise ValueError("min_sample_seconds must be positive")
        if self.max_iterations < 1:
            raise ValueError("max_iterations must be positive")


@dataclass(frozen=True)
class Measurement:
    """One median runtime measurement."""

    label: str
    size: int
    seconds_per_call: float
    iterations_per_sample: int


_BENCHMARK_SINK = 0


def _run_batch(function: WorkFunction, size: int, iterations: int) -> float:
    """Run one timed batch and preserve the computed result."""
    global _BENCHMARK_SINK

    checksum = 0
    start = perf_counter()
    for _ in range(iterations):
        checksum ^= function(size)
    elapsed = perf_counter() - start
    _BENCHMARK_SINK ^= checksum
    return elapsed


def calibrate_iterations(
    function: WorkFunction,
    size: int,
    min_sample_seconds: float,
    max_iterations: int,
) -> int:
    """Choose enough calls to make one timing sample measurable."""
    iterations = 1

    while True:
        elapsed = _run_batch(function, size, iterations)
        if elapsed >= min_sample_seconds or iterations >= max_iterations:
            return iterations
        iterations = min(iterations * 2, max_iterations)


def measure_runtime(
    function: WorkFunction,
    size: int,
    config: BenchmarkConfig,
) -> tuple[float, int]:
    """Return median seconds per call and calibrated batch size."""
    iterations = calibrate_iterations(
        function,
        size,
        config.min_sample_seconds,
        config.max_iterations,
    )

    samples = [
        _run_batch(function, size, iterations) / iterations
        for _ in range(config.repeats)
    ]
    return median(samples), iterations


def run_benchmarks(
    functions: Mapping[str, WorkFunction],
    config: BenchmarkConfig,
) -> list[Measurement]:
    """Measure every function at every configured input size."""
    config.validate()
    measurements = []

    for label, function in functions.items():
        for size in config.sizes:
            seconds_per_call, iterations = measure_runtime(function, size, config)
            measurements.append(
                Measurement(
                    label=label,
                    size=size,
                    seconds_per_call=seconds_per_call,
                    iterations_per_sample=iterations,
                )
            )

    return measurements


def estimate_log_log_slope(measurements: Sequence[Measurement]) -> float:
    """Estimate the slope of log(runtime) against log(input size)."""
    if len(measurements) < 2:
        raise ValueError("at least two measurements are required")

    x_values = [math.log(point.size) for point in measurements]
    y_values = [math.log(point.seconds_per_call) for point in measurements]
    x_mean = sum(x_values) / len(x_values)
    y_mean = sum(y_values) / len(y_values)

    numerator = sum(
        (x_value - x_mean) * (y_value - y_mean)
        for x_value, y_value in zip(x_values, y_values)
    )
    denominator = sum((x_value - x_mean) ** 2 for x_value in x_values)
    if denominator == 0:
        raise ValueError("input sizes must not all be equal")
    return numerator / denominator


def group_by_label(
    measurements: Sequence[Measurement],
) -> dict[str, list[Measurement]]:
    """Group measurements while preserving their input order."""
    grouped: dict[str, list[Measurement]] = {}
    for point in measurements:
        grouped.setdefault(point.label, []).append(point)
    return grouped


def write_csv(measurements: Sequence[Measurement], destination: Path) -> None:
    """Write raw benchmark measurements to a CSV file."""
    destination.parent.mkdir(parents=True, exist_ok=True)
    with destination.open("w", newline="", encoding="utf-8") as csv_file:
        writer = csv.writer(csv_file)
        writer.writerow(
            ["curve", "input_size", "seconds_per_call", "iterations_per_sample"]
        )
        for point in measurements:
            writer.writerow(
                [
                    point.label,
                    point.size,
                    f"{point.seconds_per_call:.12g}",
                    point.iterations_per_sample,
                ]
            )


def create_plot(measurements: Sequence[Measurement], destination: Path) -> None:
    """Create a log-log runtime plot using a non-interactive backend."""
    import matplotlib

    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    destination.parent.mkdir(parents=True, exist_ok=True)
    figure, axis = plt.subplots(figsize=(10, 6))

    for label, points in group_by_label(measurements).items():
        axis.plot(
            [point.size for point in points],
            [point.seconds_per_call for point in points],
            marker="o",
            linewidth=2,
            label=f"Curve {label}",
        )

    axis.set_xscale("log", base=2)
    axis.set_yscale("log")
    axis.set_xlabel("Input size n")
    axis.set_ylabel("Median runtime per call (seconds)")
    axis.set_title("Runtime growth of five unknown complexity classes")
    axis.grid(True, which="both", linestyle=":", alpha=0.65)
    axis.legend()
    figure.tight_layout()
    figure.savefig(destination, dpi=160)
    plt.close(figure)


def print_summary(measurements: Sequence[Measurement]) -> None:
    """Print runtimes, doubling ratios, and empirical slopes."""
    print("\nEmpirical results")
    print("=" * 76)

    for label, points in group_by_label(measurements).items():
        runtimes = [point.seconds_per_call for point in points]
        ratios = [
            current / previous
            for previous, current in zip(runtimes, runtimes[1:])
            if previous > 0
        ]
        ratio_text = ", ".join(f"{ratio:.2f}" for ratio in ratios)
        slope = estimate_log_log_slope(points)

        print(f"Curve {label}")
        print(
            "  runtimes: "
            + ", ".join(
                f"n={point.size}: {point.seconds_per_call:.3e}s"
                for point in points
            )
        )
        print(f"  doubling ratios: {ratio_text}")
        print(f"  estimated log-log slope: {slope:.3f}")


def parse_arguments() -> argparse.Namespace:
    """Parse command-line options."""
    parser = argparse.ArgumentParser(
        description="Measure five unknown runtime growth curves."
    )
    parser.add_argument(
        "--sizes",
        type=int,
        nargs="+",
        default=list(DEFAULT_SIZES),
        help="strictly increasing input sizes (default: powers of two 128..4096)",
    )
    parser.add_argument(
        "--repeats",
        type=int,
        default=5,
        help="timing samples per function and size (default: 5)",
    )
    parser.add_argument(
        "--min-sample-ms",
        type=float,
        default=10.0,
        help="target duration for one calibrated sample in ms (default: 10)",
    )
    parser.add_argument(
        "--output-dir",
        type=Path,
        default=DEFAULT_OUTPUT_DIR,
        help="directory for CSV and PNG output",
    )
    return parser.parse_args()


def main() -> None:
    """Run the benchmark lab and persist its results."""
    arguments = parse_arguments()
    config = BenchmarkConfig(
        sizes=tuple(arguments.sizes),
        repeats=arguments.repeats,
        min_sample_seconds=arguments.min_sample_ms / 1_000,
    )

    print(
        f"Measuring {len(FUNCTIONS)} curves across {len(config.sizes)} sizes. "
        "Please wait..."
    )
    measurements = run_benchmarks(FUNCTIONS, config)

    csv_path = arguments.output_dir / "measurements.csv"
    plot_path = arguments.output_dir / "runtime_growth.png"
    write_csv(measurements, csv_path)
    create_plot(measurements, plot_path)
    print_summary(measurements)

    print("\nArtifacts")
    print(f"  CSV:  {csv_path}")
    print(f"  Plot: {plot_path}")


if __name__ == "__main__":
    main()

File: DSA-lab-java/modules/01-komplexitaetsanalyse-big-o/projects/01-basic/src/main/java/lab/dsa/module01/basic/RuntimeLab.java
package lab.dsa.module01.basic;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.awt.image.BufferedImage;
import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.function.IntToLongFunction;
import javax.imageio.ImageIO;

/** Benchmarks five functions with growth from constant to quadratic. */
public final class RuntimeLab {
    public static final List<Integer> DEFAULT_SIZES =
            List.of(128, 256, 512, 1_024, 2_048, 4_096);
    public static final Path DEFAULT_OUTPUT_DIRECTORY = Paths.get("results");

    private static final Map<String, IntToLongFunction> FUNCTIONS = createFunctions();
    private static volatile long benchmarkSink;

    private RuntimeLab() {
    }

    /** Performs a fixed amount of work independently of the input size. */
    public static long curveA(int size) {
        long operations = 0;
        for (int index = 0; index < 32; index++) {
            operations++;
        }
        return operations;
    }

    /** Repeatedly halves the remaining problem size. */
    public static long curveB(int size) {
        long operations = 0;
        int remaining = size;
        while (remaining > 1) {
            for (int index = 0; index < 32; index++) {
                operations++;
            }
            remaining /= 2;
        }
        return operations;
    }

    /** Visits one item for every unit of input. */
    public static long curveC(int size) {
        long operations = 0;
        for (int item = 0; item < size; item++) {
            for (int index = 0; index < 8; index++) {
                operations++;
            }
        }
        return operations;
    }

    /** Performs a halving process for every input item. */
    public static long curveD(int size) {
        long operations = 0;
        for (int item = 0; item < size; item++) {
            int remaining = size;
            while (remaining > 1) {
                operations++;
                remaining /= 2;
            }
        }
        return operations;
    }

    /** Visits every unordered pair of distinct positions. */
    public static long curveE(int size) {
        long operations = 0;
        for (int left = 0; left < size; left++) {
            for (int right = 0; right < left; right++) {
                operations++;
            }
        }
        return operations;
    }

    /** Immutable configuration for a complete benchmark series. */
    public record BenchmarkConfig(
            List<Integer> sizes,
            int repeats,
            double minimumSampleSeconds,
            int maximumIterations) {

        public BenchmarkConfig {
            sizes = List.copyOf(sizes);
        }

        public static BenchmarkConfig defaults() {
            return new BenchmarkConfig(DEFAULT_SIZES, 5, 0.01, 1_048_576);
        }

        /** Rejects configurations that cannot produce valid benchmark data. */
        public void validate() {
            if (sizes.isEmpty()) {
                throw new IllegalArgumentException("At least one input size is required");
            }
            if (sizes.stream().anyMatch(size -> size < 2)) {
                throw new IllegalArgumentException("All input sizes must be at least 2");
            }
            for (int index = 1; index < sizes.size(); index++) {
                if (sizes.get(index - 1) >= sizes.get(index)) {
                    throw new IllegalArgumentException("Input sizes must be strictly increasing");
                }
            }
            if (repeats < 1) {
                throw new IllegalArgumentException("Repeats must be positive");
            }
            if (minimumSampleSeconds <= 0) {
                throw new IllegalArgumentException("Minimum sample duration must be positive");
            }
            if (maximumIterations < 1) {
                throw new IllegalArgumentException("Maximum iterations must be positive");
            }
        }
    }

    /** One median runtime measurement. */
    public record Measurement(
            String label,
            int size,
            double secondsPerCall,
            int iterationsPerSample) {
    }

    private record RuntimeResult(double secondsPerCall, int iterations) {
    }

    private record Arguments(List<Integer> sizes, int repeats, double minimumSampleSeconds,
                             Path outputDirectory, boolean help) {
    }

    private static Map<String, IntToLongFunction> createFunctions() {
        Map<String, IntToLongFunction> functions = new LinkedHashMap<>();
        functions.put("A", RuntimeLab::curveA);
        functions.put("B", RuntimeLab::curveB);
        functions.put("C", RuntimeLab::curveC);
        functions.put("D", RuntimeLab::curveD);
        functions.put("E", RuntimeLab::curveE);
        return java.util.Collections.unmodifiableMap(functions);
    }

    private static double runBatch(IntToLongFunction function, int size, int iterations) {
        long checksum = 0;
        long start = System.nanoTime();
        for (int iteration = 0; iteration < iterations; iteration++) {
            checksum ^= function.applyAsLong(size);
        }
        double elapsed = (System.nanoTime() - start) / 1_000_000_000.0;
        benchmarkSink ^= checksum;
        return elapsed;
    }

    public static int calibrateIterations(
            IntToLongFunction function,
            int size,
            double minimumSampleSeconds,
            int maximumIterations) {
        int iterations = 1;
        while (true) {
            double elapsed = runBatch(function, size, iterations);
            if (elapsed >= minimumSampleSeconds || iterations >= maximumIterations) {
                return iterations;
            }
            iterations = Math.min(iterations * 2, maximumIterations);
        }
    }

    private static RuntimeResult measureRuntime(
            IntToLongFunction function, int size, BenchmarkConfig config) {
        int iterations = calibrateIterations(
                function, size, config.minimumSampleSeconds(), config.maximumIterations());
        double[] samples = new double[config.repeats()];
        for (int repeat = 0; repeat < config.repeats(); repeat++) {
            samples[repeat] = runBatch(function, size, iterations) / iterations;
        }
        Arrays.sort(samples);
        double median = samples.length % 2 == 1
                ? samples[samples.length / 2]
                : (samples[samples.length / 2 - 1] + samples[samples.length / 2]) / 2;
        return new RuntimeResult(median, iterations);
    }

    public static List<Measurement> runBenchmarks(
            Map<String, IntToLongFunction> functions, BenchmarkConfig config) {
        config.validate();
        List<Measurement> measurements = new ArrayList<>();
        functions.forEach((label, function) -> {
            for (int size : config.sizes()) {
                RuntimeResult result = measureRuntime(function, size, config);
                measurements.add(new Measurement(
                        label, size, result.secondsPerCall(), result.iterations()));
            }
        });
        return measurements;
    }

    public static double estimateLogLogSlope(List<Measurement> measurements) {
        if (measurements.size() < 2) {
            throw new IllegalArgumentException("At least two measurements are required");
        }
        double xMean = measurements.stream().mapToDouble(point -> Math.log(point.size())).average().orElseThrow();
        double yMean = measurements.stream().mapToDouble(point -> Math.log(point.secondsPerCall())).average().orElseThrow();
        double numerator = 0;
        double denominator = 0;
        for (Measurement point : measurements) {
            double xDifference = Math.log(point.size()) - xMean;
            numerator += xDifference * (Math.log(point.secondsPerCall()) - yMean);
            denominator += xDifference * xDifference;
        }
        if (denominator == 0) {
            throw new IllegalArgumentException("Input sizes must not all be equal");
        }
        return numerator / denominator;
    }

    public static Map<String, List<Measurement>> groupByLabel(List<Measurement> measurements) {
        Map<String, List<Measurement>> grouped = new LinkedHashMap<>();
        for (Measurement measurement : measurements) {
            grouped.computeIfAbsent(measurement.label(), ignored -> new ArrayList<>()).add(measurement);
        }
        return grouped;
    }

    public static void writeCsv(List<Measurement> measurements, Path destination) throws IOException {
        Path parent = destination.getParent();
        if (parent != null) {
            Files.createDirectories(parent);
        }
        try (BufferedWriter writer = Files.newBufferedWriter(destination, StandardCharsets.UTF_8)) {
            writer.write("curve,input_size,seconds_per_call,iterations_per_sample");
            writer.newLine();
            for (Measurement point : measurements) {
                writer.write(String.format(Locale.ROOT, "%s,%d,%.12g,%d",
                        point.label(), point.size(), point.secondsPerCall(), point.iterationsPerSample()));
                writer.newLine();
            }
        }
    }

    /** Creates a dependency-free PNG chart with base-2 logarithmic axes. */
    public static void createPlot(List<Measurement> measurements, Path destination) throws IOException {
        int width = 1_200;
        int height = 720;
        int left = 105;
        int right = 55;
        int top = 70;
        int bottom = 100;
        BufferedImage image = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);
        Graphics2D graphics = image.createGraphics();
        graphics.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
        graphics.setColor(Color.WHITE);
        graphics.fillRect(0, 0, width, height);

        double minimumX = measurements.stream().mapToDouble(point -> log2(point.size())).min().orElseThrow();
        double maximumX = measurements.stream().mapToDouble(point -> log2(point.size())).max().orElseThrow();
        double minimumY = measurements.stream().mapToDouble(point -> Math.log10(point.secondsPerCall())).min().orElseThrow();
        double maximumY = measurements.stream().mapToDouble(point -> Math.log10(point.secondsPerCall())).max().orElseThrow();
        if (minimumX == maximumX) {
            maximumX = minimumX + 1;
        }
        if (minimumY == maximumY) {
            maximumY = minimumY + 1;
        }

        graphics.setColor(new Color(225, 225, 225));
        for (int line = 0; line <= 5; line++) {
            int y = top + line * (height - top - bottom) / 5;
            graphics.drawLine(left, y, width - right, y);
        }
        graphics.setColor(Color.BLACK);
        graphics.setStroke(new BasicStroke(2));
        graphics.drawLine(left, top, left, height - bottom);
        graphics.drawLine(left, height - bottom, width - right, height - bottom);
        graphics.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 24));
        graphics.drawString("Runtime growth of five unknown complexity classes", 250, 38);
        graphics.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 16));
        graphics.drawString("Input size n (log₂)", 520, height - 40);
        graphics.rotate(-Math.PI / 2);
        graphics.drawString("Median runtime per call (log₁₀ seconds)", -500, 30);
        graphics.rotate(Math.PI / 2);

        Color[] colors = {new Color(31, 119, 180), new Color(255, 127, 14),
                new Color(44, 160, 44), new Color(214, 39, 40), new Color(148, 103, 189)};
        int colorIndex = 0;
        for (Map.Entry<String, List<Measurement>> entry : groupByLabel(measurements).entrySet()) {
            graphics.setColor(colors[colorIndex % colors.length]);
            graphics.setStroke(new BasicStroke(3));
            int previousX = -1;
            int previousY = -1;
            for (Measurement point : entry.getValue()) {
                int x = left + (int) ((log2(point.size()) - minimumX) / (maximumX - minimumX)
                        * (width - left - right));
                int y = height - bottom - (int) ((Math.log10(point.secondsPerCall()) - minimumY)
                        / (maximumY - minimumY) * (height - top - bottom));
                if (previousX >= 0) {
                    graphics.drawLine(previousX, previousY, x, y);
                }
                graphics.fillOval(x - 5, y - 5, 10, 10);
                previousX = x;
                previousY = y;
            }
            graphics.drawString("Curve " + entry.getKey(), width - 145, top + 25 * colorIndex);
            colorIndex++;
        }
        graphics.dispose();
        Path parent = destination.getParent();
        if (parent != null) {
            Files.createDirectories(parent);
        }
        ImageIO.write(image, "png", destination.toFile());
    }

    private static double log2(double value) {
        return Math.log(value) / Math.log(2);
    }

    public static void printSummary(List<Measurement> measurements) {
        System.out.println("\nEmpirical results");
        System.out.println("=".repeat(76));
        groupByLabel(measurements).forEach((label, points) -> {
            List<String> runtimes = points.stream()
                    .map(point -> String.format(Locale.ROOT, "n=%d: %.3es", point.size(), point.secondsPerCall()))
                    .toList();
            List<String> ratios = new ArrayList<>();
            for (int index = 1; index < points.size(); index++) {
                ratios.add(String.format(Locale.ROOT, "%.2f",
                        points.get(index).secondsPerCall() / points.get(index - 1).secondsPerCall()));
            }
            System.out.println("Curve " + label);
            System.out.println("  runtimes: " + String.join(", ", runtimes));
            System.out.println("  doubling ratios: " + String.join(", ", ratios));
            System.out.printf(Locale.ROOT, "  estimated log-log slope: %.3f%n", estimateLogLogSlope(points));
        });
    }

    private static Arguments parseArguments(String[] arguments) {
        List<Integer> sizes = new ArrayList<>(DEFAULT_SIZES);
        int repeats = 5;
        double minimumSampleSeconds = 0.01;
        Path outputDirectory = DEFAULT_OUTPUT_DIRECTORY;
        for (int index = 0; index < arguments.length; index++) {
            switch (arguments[index]) {
                case "--help" -> {
                    return new Arguments(sizes, repeats, minimumSampleSeconds, outputDirectory, true);
                }
                case "--repeats" -> repeats = Integer.parseInt(requireValue(arguments, ++index, "--repeats"));
                case "--min-sample-ms" -> minimumSampleSeconds =
                        Double.parseDouble(requireValue(arguments, ++index, "--min-sample-ms")) / 1_000;
                case "--output-dir" -> outputDirectory = Paths.get(requireValue(arguments, ++index, "--output-dir"));
                case "--sizes" -> {
                    sizes = new ArrayList<>();
                    while (index + 1 < arguments.length && !arguments[index + 1].startsWith("--")) {
                        sizes.add(Integer.parseInt(arguments[++index]));
                    }
                }
                default -> throw new IllegalArgumentException("Unknown option: " + arguments[index]);
            }
        }
        return new Arguments(sizes, repeats, minimumSampleSeconds, outputDirectory, false);
    }

    private static String requireValue(String[] arguments, int index, String option) {
        if (index >= arguments.length) {
            throw new IllegalArgumentException("Missing value for " + option);
        }
        return arguments[index];
    }

    private static void printHelp() {
        System.out.println("Usage: mvn exec:java -Dexec.args=\"[options]\"");
        System.out.println("  --sizes N...          Strictly increasing sizes (default: 128..4096)");
        System.out.println("  --repeats N           Samples per function and size (default: 5)");
        System.out.println("  --min-sample-ms N     Target sample duration in milliseconds (default: 10)");
        System.out.println("  --output-dir PATH     Directory for CSV and PNG output (default: results)");
    }

    public static void main(String[] commandLineArguments) throws IOException {
        Arguments arguments = parseArguments(commandLineArguments);
        if (arguments.help()) {
            printHelp();
            return;
        }
        BenchmarkConfig config = new BenchmarkConfig(
                arguments.sizes(), arguments.repeats(), arguments.minimumSampleSeconds(), 1_048_576);
        System.out.printf("Measuring %d curves across %d sizes. Please wait...%n",
                FUNCTIONS.size(), config.sizes().size());
        List<Measurement> measurements = runBenchmarks(FUNCTIONS, config);
        Path csvPath = arguments.outputDirectory().resolve("measurements.csv");
        Path plotPath = arguments.outputDirectory().resolve("runtime_growth.png");
        writeCsv(measurements, csvPath);
        createPlot(measurements, plotPath);
        printSummary(measurements);
        System.out.println("\nArtifacts");
        System.out.println("  CSV:  " + csvPath);
        System.out.println("  Plot: " + plotPath);
    }
}
